### ISSUE TITLE
[APP SUBMITTED]: [Errno 2] No such file or directory: '/storage/media/media/downloads/series/Madam.Secretary.S01E08.720p.HDTV.X264-DIMENSION.mkv'

### ISSUE THREAD

**Issue by konomimaster** [SiCKRAGETV/sickrage-issues#1334]
### INFO
Python Version: **2.7.3 (default, Mar 31 2015, 15:16:49) [GCC 4.9.2]**
Operating System: **Linux-3.18.10-armv7l-with-glibc2.4**
Locale: unknown
Branch: **master**
Commit: SiCKRAGETV/SickRage@281409d366b228a455879e02b2d55d3967db29e4
Link to Log: https://gist.github.com/4286325ce56a404f5c36
### ERROR
```
Thread-70 :: Failed doing webui callback: Traceback (most recent call last):
  File "/storage/.kodi/addons/service.downloadmanager.SickPotatoHead/resources/SickBeard/sickbeard/webserve.py", line 258, in async_call
    result = function(**kwargs)
  File "/storage/.kodi/addons/service.downloadmanager.SickPotatoHead/resources/SickBeard/sickbeard/webserve.py", line 2144, in processEpisode
    is_priority=is_priority, delete_on=delete_on, failed=failed, type=type)
  File "/storage/.kodi/addons/service.downloadmanager.SickPotatoHead/resources/SickBeard/sickbeard/processTV.py", line 203, in processDir
    process_media(path, [video], nzbName, process_method, force, is_priority, result)
  File "/storage/.kodi/addons/service.downloadmanager.SickPotatoHead/resources/SickBeard/sickbeard/processTV.py", line 504, in process_media
    result.result = processor.process()
  File "/storage/.kodi/addons/service.downloadmanager.SickPotatoHead/resources/SickBeard/sickbeard/postProcessor.py", line 866, in process
    existing_file_status = self._checkForExistingFile(ep_obj.location)
  File "/storage/.kodi/addons/service.downloadmanager.SickPotatoHead/resources/SickBeard/sickbeard/postProcessor.py", line 132, in _checkForExistingFile
    if ek.ek(os.path.getsize, existing_file) > ek.ek(os.path.getsize, self.file_path):
  File "/storage/.kodi/addons/service.downloadmanager.SickPotatoHead/resources/SickBeard/sickbeard/encodingKludge.py", line 58, in ek
    result = func(*[ss(x) if isinstance(x, (str, unicode)) else x for x in args], **kwargs)
  File "/home/stephan/projects/openelec-5.0/build.OpenELEC-RPi2.arm-5.0.8/Python-2.7.3/.install_pkg/usr/lib/python2.7/genericpath.py", line 49, in getsize
OSError: [Errno 2] No such file or directory: '/storage/media/media/downloads/series/Madam.Secretary.S01E08.720p.HDTV.X264-DIMENSION.mkv'

```
---
_STAFF NOTIFIED_: @SiCKRAGETV/owners @SiCKRAGETV/moderators

**Comment by fernandog** [comment]
@tagforce we are getting some of this issues. Confirmed issue?

**Comment by TagForce** [comment]
@fernandog I, personally, have not seen this issue. If multiple users are reporting this, then something is going on... I'll have a look as soon as I can.

**Comment by TagForce** [comment]
@konomimaster @fernandog 
Ok, that's strange...

From the log:
```
2015-04-20 08:36:42 DEBUG    Thread-70 :: PostProcessing Files: ['Madam.Secretary.S01E08.720p.HDTV.X264-DIMENSION.mkv', 'Madam.Secretary.S01E10.Collateral.Damage.1080p.WEB-DL.DD5.1.H.264-BS.mkv', 'Madam.Secretary.S01E10.Collateral.Damage.1080p.WEB-DL.DD5.1.H.264-BS.srt', 'Madam.Secretary.S01E14.Whisper.of.the.Ax.1080p.WEB-DL.DD5.1.H.264-BS.mkv', 'Madam.Secretary.S01E14.Whisper.of.the.Ax.1080p.WEB-DL.DD5.1.H.264-BS.srt', 'ugly.betty.S04102.hdtv.xvid-notv.avi']
2015-04-20 08:36:42 DEBUG    Thread-70 :: PostProcessing Dirs: ['Suits S04E11 (2014) H.264 (WEB-DL) DD5.1 1080p', '.sync']
2015-04-20 08:36:42 DEBUG    Thread-70 :: PostProcessing Path: /storage/media/media/downloads/series
2015-04-20 08:36:42 DEBUG    Thread-65 :: Moving file from /storage/media/media/downloads/series/Madam.Secretary.S01E08.720p.HDTV.X264-DIMENSION.mkv to /storage/media/media/serier/Madam Secretary/Madam Secretary - S01E08 - Need to Know.mkv
2015-04-20 08:36:42 DEBUG    Thread-65 :: Checking if the path /storage/media/media/serier/Madam Secretary already exists
2015-04-20 08:36:42 DEBUG    Thread-65 :: Destination folder for this episode: /storage/media/media/serier/Madam Secretary
2015-04-20 08:36:42 DEBUG    Thread-65 :: formatting pattern: %SN - S%0SE%0E - %EN -> Madam Secretary - S01E08 - Need to Know
2015-04-20 08:36:42 DEBUG    Thread-65 :: Parsed Madam.Secretary.S01E08.720p.HDTV.X264-DIMENSION into Madam Secretary - S1E8 [ANIME VER: -1] [GROUP: DIMENSION] [ABD: False] [ANIME: False] [whichReg: ['standard']]
2015-04-20 08:36:42 DEBUG    Thread-65 :: Using NORMAL regexs
2015-04-20 08:36:42 DEBUG    Thread-65 :: Parsed /storage/media/media/serier/Madam Secretary/Madam Secretary - S01E08 - Need to Know into Madam Secretary - S1E8 [ANIME VER: -1] [ABD: False] [ANIME: False] [whichReg: ['standard']]
2015-04-20 08:36:42 DEBUG    Thread-70 :: TV_DOWNLOAD_DIR: /storage/media/media/downloads/series
2015-04-20 08:36:42 DEBUG    Thread-70 :: Processing folder /storage/media/media/downloads/series
2015-04-20 08:36:42 DEBUG    Thread-65 :: Using NORMAL regexs
2015-04-20 08:36:42 INFO     Thread-65 :: Folder name (series) appears to be a valid release name. Using it.
2015-04-20 08:36:42 DEBUG    Thread-65 :: Using ALL regexes
2015-04-20 08:36:42 DEBUG    Thread-65 :: Found release name Madam.Secretary.S01E08.720p.HDTV.X264-DIMENSION
2015-04-20 08:36:42 INFO     Thread-65 :: Trying to clean any empty folders under /storage/media/media/serier/Madam Secretary
2015-04-20 08:36:42 DEBUG    Thread-65 :: Deleting file /storage/media/media/serier/Madam Secretary/Madam Secretary - S01E08 - Need to Know.en.srt
2015-04-20 08:36:42 DEBUG    Thread-65 :: Deleting file /storage/media/media/serier/Madam Secretary/Madam Secretary - S01E08 - Need to Know.mkv
2015-04-20 08:36:40 DEBUG    Thread-65 :: Trying to determine free space on destination drive
2015-04-20 08:36:40 DEBUG    Thread-65 :: This download is marked a priority download so I'm going to replace an existing file if I find one
2015-04-20 08:36:40 DEBUG    Thread-65 :: File /storage/media/media/serier/Madam Secretary/Madam Secretary - S01E08 - Need to Know.mkv is the same size as /storage/media/media/downloads/series/Madam.Secretary.S01E08.720p.HDTV.X264-DIMENSION.mkv
```

Read that down up (log is reversed, latest entries are on top).
Notice the thread names:
It starts off PP-ing the file (thread-65)...
Then right at the moment that thread is preparing the destination folder for moving the file, a new thread starts PP-ing the same files (thread-70)...
So once thread-65 is done moving the file, it gets deleted, and at that time thread-70 starts preparing the destination for moving it, and then fails because thread-65 has just moved it...

I can't for the life of me reproduce this error, because Post Processing on my systems (both production and development machines) will only start when the last PP run has finished (I've shown you logs that prove that a while ago).

**Comment by fernandog** [comment]
@tagforce where in the code it checks if if last PP already finished?

Sorry about that hotfix and url error codes

**Comment by TagForce** [comment]
@fernandog It doesn't... At least, I couldn't find it, but it won't start a new thread if the old one hasn't yet finished, at least on both my machines.

**Comment by fernandog** [comment]
@tagforce Maybe it runs really fast and just few episodes to PP. I still think we should add the check if is running

**Comment by abeloin** [comment]
@fernandog The scheduler cannot run 2 PP at the same time, see scheduler.py.

 Is it possible to run 2 PP if you run a manual PP at the same time or if an external script start it?

**Comment by TagForce** [comment]
@abeloin @fernandog 
There's nothing preventing it from happening when you directly call processDir()...
processDir() is called from the following 3 locations:
autoPostProcesser.py - The Scheduler
webapi.py - from CMD_PostProcess(ApiCall)
webserve.py - from HomePostProcess(Home)

So the last 2 might actually be run simultaneously...

If we don't want multiple PP threads to run, does PP need to be thread safe? Because if not, we could just abuse a global for a quick and dirty fix while we fix it properly. Something like sickbeard.PPactive
and then check if sickbeard.PPactive in the first line of processDir, log and break if it is. Then set it to True at the next line, and set it to false again at each and every return in processDir.

**Comment by fernandog** [comment]
@TagForce @abeloin what is the best long term solution?

**Comment by abeloin** [comment]
@TagForce @fernandog I think we should implement a PP queue like the search and show queue. We set the auto PP with a low priority and the manual one(webapi and webserve) with High. For the returned results, we use a callback or yield the results.

**Comment by fernandog** [comment]
@SiCKRAGETV/moderators agree

also implement this:
https://github.com/SiCKRAGETV/sickrage-issues/issues/1260

**Comment by TagForce** [comment]
@abeloin Sounds like a plan.

**Comment by miigotu** [comment]
+1 @abeloin

**Comment by OmgImAlexis** [comment]
Related https://github.com/SiCKRAGETV/sickrage-issues/issues/2543

**Comment by fernandog** [comment]
@MGaetan89 
with the union of webserve code + webapi code (is that your goal?) we can handle better the manual API PP with the manual UI PP and etc

**Comment by MGaetan89** [comment]
Yes both will run the same code, so it will be easier to queue PP and prevent multiple instances at the same time.
